server/src/main/java/org/elasticsearch/common/RunningTimeRecorder.java

```diff
@@ -0,0 +1,120 @@
+ * more frequent than the running operations.
+ */
+public class RunningTimeRecorder {
+    public static final RunningTimeRecorder NO_OP = new RunningTimeRecorder(() -> 0L) {
+        private static final Releasable EMPTY_RELEASABLE = () -> {};
+
+        @Override
+        public Releasable trackRunningTime() {
+            return EMPTY_RELEASABLE;
+        }
+
+        @Override
+        public long totalRunningTimeInNanos() {
+            return 0L;
+        }
+    };
+
+    private final LongSupplier relativeTimeSupplier;
+    private final LongAdder totalRunningTimeInNanos = new LongAdder();
+
+    private final Set<OperationRunningTimeRecorder> operationRunningTimeRecorders = ConcurrentCollections.newConcurrentSet();
+
+    public RunningTimeRecorder(LongSupplier relativeTimeSupplier) {
+        this.relativeTimeSupplier = relativeTimeSupplier;
+    }
+
+    /**
+     * Starts recording the running time of an operation, the recording stops once the returned {@link Releasable} is
+     * closed.
+     * @return a {@link Releasable} that should be called once the operation is finished to stop the running time recording.
+     */
+    public Releasable trackRunningTime() {
+        final var recorder = new OperationRunningTimeRecorder(getCurrentTimeInNanos());
+        operationRunningTimeRecorders.add(recorder);
+        return recorder;
+    }
+
+    private void onOperationComplete(OperationRunningTimeRecorder tracker, long runningTimeInNanos) {
+        boolean removed = operationRunningTimeRecorders.remove(tracker);
+        if (removed) {
+            totalRunningTimeInNanos.add(runningTimeInNanos);
+        }
+    }
+
+    private class OperationRunningTimeRecorder implements Releasable {
+        private static final long FINISHED_OP_SENTINEL = Long.MIN_VALUE;
+
+        private final AtomicLong latestReadingRelativeTime;
+
+        private OperationRunningTimeRecorder(long relativeStartTime) {
+            this.latestReadingRelativeTime = new AtomicLong(relativeStartTime);
+        }
+
+        @Override
+        public void close() {
+            final long previousReadingTime = latestReadingRelativeTime.getAndSet(FINISHED_OP_SENTINEL);
+            assert previousReadingTime != FINISHED_OP_SENTINEL : previousReadingTime;
+
+            onOperationComplete(this, getCurrentTimeInNanos() - previousReadingTime);
+        }
+
+        private long getPartialRunningTimeAndUpdateLatestReadingTime() {
+            long previousReadingTime;
+            long updatedReadingTime;
+            do {
+                previousReadingTime = latestReadingRelativeTime.get();
+                updatedReadingTime = getCurrentTimeInNanos();
+            } while (previousReadingTime != FINISHED_OP_SENTINEL
+                && latestReadingRelativeTime.compareAndSet(previousReadingTime, updatedReadingTime) == false);
+            return previousReadingTime == FINISHED_OP_SENTINEL ? 0 : updatedReadingTime - previousReadingTime;
```

Comment:
Since we only need to total indexing time + the total duration of the shard, I think we can get away with not resetting any state here and not returning the delta?

At rollover time, we will simply ask for the the total indexing time + total duration of the shard.

I think that would simplify the tracking quite a bit?